### PR TITLE
Repo API logic (Single API)

### PR DESCRIPTION
**Pull request by Xorima**
# Description

This commit will setup the logic required for repo creation
it adds in the single API that supports the 4 types of repositories we have

## Issues Resolved

List any existing issues this PR resolves

## Check List

- [ ] All tests pass. See TESTING.md for details.
- [ ] New functionality includes testing.
- [ ] New functionality has been documented in the README if applicable.

### CODE REVIEW

**Comment by Xorima** [comment]
@damacus

**Comment by damacus** [comment]
thanks @Xorima I'll take a look at this first thing tomorrow morning 👍🏼

File: .github/workflows/terraform-lint.yml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+jobs:
+  terraform-lint:
+
+    runs-on: ubuntu-latest
+
+    steps:
+    - name: Check out code
+      uses: actions/checkout@master
+    - name: Lint Terraform
+      uses: actionshub/terraform-lint@main
```
Comment (damacus) [inline .github/workflows/terraform-lint.yml:14]: One for the future, but I wonder if it's worth looking at using [the official Actions instead](https://github.com/hashicorp/setup-terraform)?

File: main.tf [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+  source                    = "./modules/repository"
+  name                      = each.value.name
+  repo_type                 = each.value.repo_type
+  supermarket_name_override = each.value.supermarket_name_override
+  description_override      = each.value.description_override
+  homepage_url_override     = each.value.homepage_url_override
+  additional_topics         = each.value.additional_topics
+  additional_status_checks  = each.value.additional_status_checks != null ? each.value.additional_status_checks : []
+  projects_enabled          = each.value.projects_enabled
+}
```
Comment (damacus) [inline main.tf:12]: This looks pretty good, do we need to do that null check for any of the other values?

File: modules/repository/main.tf [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  ]
+
+  required_status_checks {
+    strict   = true
+    contexts = local.status_checks
+  }
+
+  required_pull_request_reviews {
+    dismiss_stale_reviews      = true
+    require_code_owner_reviews = false
```
Comment (damacus) [inline modules/repository/main.tf:52]: Does hit actually want to be true here?

File: terraform.tfvars.json [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    },
+    {
+      "name": "tf-test-with-url",
+      "repo_type": "terraform",
+      "additional_topics": ["jason-testing"],
+      "homepage_url_override": "google.com"
+    },
+    {
+      "name": "ide-test-with-url",
+      "repo_type": "ide",
```
Comment (damacus) [inline terraform.tfvars.json:26]:
```suggestion
      "type": "ide",
```

I think just "type" works here?

**Comment by damacus** [comment]
We might also want some outputs for other consumers: 

```HCL
output "cookbooks" {
  value = [for x in var.repository: x.name if x.repo_type == "cookbook"]
}

output "terraform" {
  value = [for x in var.repository: x.name if x.repo_type == "terraform"]
}

output "ide" {
  value = [for x in var.repository: x.name if x.repo_type == "ide"]
}
```

With your example `terraform.tfvars.json`

```bash
Outputs:

cookbooks = [
  "vscode",
  "foobar",
]
ide = [
  "ide-test-with-url",
]
terraform = [
  "tf-test",
  "tf-test-with-url",
]
```

File: main.tf (hunk 2 above)
Comment (Xorima) [reply main.tf:12]: Nope, as the others come through properly a list cant be passed as null but a string can.

File: modules/repository/main.tf (hunk 3 above)
Comment (Xorima) [reply modules/repository/main.tf:52]: We can do but we haven't setup codeowners yet?

File: terraform.tfvars.json (hunk 4 above)
Comment (Xorima) [reply terraform.tfvars.json:26]: I prefer a more specific name as if a type property ever comes out of github I don't want a clash

**Pull request merged**